zeusd/src/devices/cpu/linux.rs

```diff
@@ -0,0 +1,219 @@
+//! CPU power measurement with RAPL. Only supported on Linux.
+
+use std::fs;
+use std::io::Read;
+use std::path::PathBuf;
+use std::string::String;
+use std::sync::{Arc, RwLock};
+
+use once_cell::sync::OnceCell;
+use tokio::io::AsyncReadExt;
+use tokio::task::JoinHandle;
+use tokio::time::{sleep, Duration};
+
+use crate::devices::cpu::{CpuManager, PackageInfo};
+use crate::error::ZeusdError;
+
+static RAPL_DIR: &'static str = "/sys/class/powercap/intel-rapl";
+
+pub struct RaplCpu {
+    cpu: Arc<PackageInfo>,
+    dram: Arc<Option<PackageInfo>>,
+    cpu_monitoring_task: OnceCell<JoinHandle<Result<(), ZeusdError>>>,
+    dram_monitoring_task: OnceCell<JoinHandle<Result<(), ZeusdError>>>,
+}
+
+impl RaplCpu {
+    pub fn init(_index: u32) -> Result<Self, ZeusdError> {
+        let fields = RaplCpu::get_available_fields(_index)?;
+        let cpu = Arc::new(fields.0);
+        let dram = Arc::new(fields.1);
+        Ok(Self {
+            cpu,
+            dram,
+            cpu_monitoring_task: OnceCell::new(),
+            dram_monitoring_task: OnceCell::new(),
+        })
+    }
+}
+
+impl PackageInfo {
```

Comment:
Yup, the `new` method is called in `get_available_fields` and it was just to separate the logic of getting all the fields of `PackageInfo` and finding out all the available packages to avoid code duplication for making a package. In `macos` I just, hardcode the individual fields of `PackageInfo`.